File: src/cli-sample/PrintDetailed.cs
using System;

namespace Xt
{
    public static class PrintDetailed
    {
        static void OnError(string location, string message) 
        => Console.WriteLine($"{location}: {message}");

        public static void Main()
        {
            using IDisposable audio = XtAudio.Init("Sample", IntPtr.Zero, OnError);
            try
            {
                XtVersion version = XtAudio.GetVersion();
                Console.WriteLine("Version: " + version.major + "." + version.minor);
                XtSystem pro = XtAudio.SetupToSystem(XtSetup.ProAudio);
                Console.WriteLine("Pro Audio: " + pro + " (" + (XtAudio.GetService(pro) != null) + ")");
                XtSystem system = XtAudio.SetupToSystem(XtSetup.SystemAudio);
                Console.WriteLine("System Audio: " + system + " (" + (XtAudio.GetService(system) != null) + ")");
                XtSystem consumer = XtAudio.SetupToSystem(XtSetup.ConsumerAudio);
                Console.WriteLine("Consumer Audio: " + consumer + " (" + (XtAudio.GetService(consumer) != null) + ")");

                foreach (XtSystem s in XtAudio.GetSystems())
                {
                    XtService service = XtAudio.GetService(s);
                    Console.WriteLine("System: " + s);
                    Console.WriteLine("  Device count: " + service.GetDeviceCount());
                    Console.WriteLine("  Capabilities: " + service.GetCapabilities());
                    using XtDevice defaultInput = service.OpenDefaultDevice(false);
                    if (defaultInput != null) Console.WriteLine("  Default input: " + defaultInput);
                    using XtDevice defaultOutput = service.OpenDefaultDevice(true);
                    if (defaultOutput != null) Console.WriteLine("  Default output: " + defaultOutput);

                    for (int d = 0; d < service.GetDeviceCount(); d++)
                        using (XtDevice device = service.OpenDevice(d))
                        {
                            XtMix? mix = device.GetMix();
                            Console.WriteLine("  Device " + device + ":");
                            Console.WriteLine("    Input channels: " + device.GetChannelCount(false));
                            Console.WriteLine("    Output channels: " + device.GetChannelCount(true));
                            Console.WriteLine("    Interleaved access: " + device.SupportsAccess(true));
                            Console.WriteLine("    Non-interleaved access: " + device.SupportsAccess(false));
                            if (mix != null) Console.WriteLine("    Current mix: " + mix.Value.rate + " " + mix.Value.sample);
                        }
                }
            } catch (XtException e)
            {
                Console.WriteLine(XtAudio.GetErrorInfo(e.GetError()));
            }
        }
    }
}

File: src/java-sample/com/xtaudio/xt/sample/PrintDetailed.java
package com.xtaudio.xt.sample;

import java.util.Optional;
import com.xtaudio.xt.*;
import static com.xtaudio.xt.NativeTypes.*;

public class PrintDetailed {

    static void onError(String location, String message) {
        System.out.println(location + ": " + message);
    }

    public static void main() {
        try(XtAudio audio = new XtAudio("Sample", null, PrintDetailed::onError)) {
            XtVersion version = XtAudio.getVersion();
            System.out.println("Version: " + version.major + "." + version.minor);
            XtSystem pro = XtAudio.setupToSystem(XtSetup.PRO_AUDIO);
            System.out.println("Pro Audio: " + pro + " (" + (XtAudio.getService(pro) != null) + ")");
            XtSystem system = XtAudio.setupToSystem(XtSetup.SYSTEM_AUDIO);
            System.out.println("System Audio: " + system + " (" + (XtAudio.getService(system) != null) + ")");
            XtSystem consumer = XtAudio.setupToSystem(XtSetup.CONSUMER_AUDIO);
            System.out.println("Consumer Audio: " + consumer + " (" + (XtAudio.getService(consumer) != null) + ")");

            for(XtSystem s: XtAudio.getSystems()) {
                XtService service = XtAudio.getService(s);
                System.out.println("System " + s + ":");
                System.out.println("  Device count: " + service.getDeviceCount());
                System.out.println("  Capabilities: " + service.getCapabilities());
                try(XtDevice defaultInput = service.openDefaultDevice(false)) {
                    System.out.println("  Default input: " + defaultInput);
                }
                try(XtDevice defaultOutput = service.openDefaultDevice(true)) {
                    System.out.println("  Default output: " + defaultOutput);
                }

                for(int d = 0; d < service.getDeviceCount(); d++)
                    try(XtDevice device = service.openDevice(d)) {
                        Optional<XtMix> mix = device.getMix();
                        System.out.println("  Device " + device + ":");
                        System.out.println("    Input channels: " + device.getChannelCount(false));
                        System.out.println("    Output channels: " + device.getChannelCount(true));
                        System.out.println("    Interleaved access: " + device.supportsAccess(true));
                        System.out.println("    Non-interleaved access: " + device.supportsAccess(false));
                        if(mix.isPresent())
                            System.out.println("    Current mix: " + mix.get().rate + " " + mix.get().sample);
                    }
            }
        } catch(XtException e) {
            System.out.println(XtAudio.getErrorInfo(e.getError()));
        }
    }
}